src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+
+        self.output_hidden_states = config.output_hidden_states
+
+        self.dropout = config.dropout
+        embed_dim = embed_tokens.embedding_dim
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_source_positions = config.max_position_embeddings
+
+        self.embed_tokens = embed_tokens
+        self.embed_scale = None
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings+1+self.padding_idx, embed_dim, self.padding_idx)
+
+        self.layers = nn.ModuleList([])
+        self.layers.extend([
+            EncoderLayer(config)
+            for i in range(config.encoder_layers)
+        ])
+        self.emb_layer_norm = LayerNorm(embed_dim)
+
+    def forward(self, input_ids, attention_mask=None, output_attentions=False):
+        """
+
+        Args:
+            input_ids (LongTensor): tokens in the source language of shape
+                `(batch, src_len)`
+            attention_mask (torch.LongTensor): indicating which indices are padding tokens.
+        Returns:
+            Tuple comprised of:
+                - **x** (Tensor): the last encoder layer's output of
+                  shape `(src_len, batch, embed_dim)`
+                - **encoder_states** (List[Tensor]): all intermediate
+                  hidden states of shape `(src_len, batch, embed_dim)`.
+                  Only populated if *output_hidden_states:* is True.
+                - **all_attentions** (List[Tensor]): Attention weights for each layer.
+                During training might not be of length n_layers because of layer dropout.
+
+        """
+
+        # remove bos to be consistent with fairseq version
+        input_ids = input_ids[:, 1:]
+        if attention_mask is not None:
+            attention_mask = attention_mask[:, 1:]
+
+        if attention_mask is not None:
+            attention_mask = invert_mask(attention_mask)
+
+        inputs_embeds =self.embed_tokens(input_ids)
+        embed_pos, real_positions = self.embed_positions(input_ids)
+        x = inputs_embeds + embed_pos
+        x = self.emb_layer_norm(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        # B x T x C -> T x B x C
+        x = x.transpose(0, 1)
+
+        encoder_states, all_attentions = [], []
+        for encoder_layer in self.layers:
+            if self.output_hidden_states:
+                encoder_states.append(x)
+            x, attn = encoder_layer(x, attention_mask, output_attentions=output_attentions)
+            if output_attentions:
+                all_attentions.append(attn)
+
+        if self.output_hidden_states:
+            encoder_states.append(x)
+
+        # T x B x C -> B x T x C
+        encoder_states = [hidden_state.transpose(0, 1) for hidden_state in encoder_states]
+        x = x.transpose(0, 1)
+        results = (x, encoder_states, all_attentions)
+        return results
+
+    def ngram_attention_bias(length, num_skip):
+        bias_result = []
```

Comment:
a comment would be great here

Reply:
Is there a way to not use two `for` loops here? This function is called at every forward step and `for` loops render GPU very slow in my experience